relayer/relays/parachain/ethereum-writer.go

```diff
@@ -283,6 +296,26 @@ func (wr *EthereumWriter) WriteBasicChannel(
 	return nil
 }
 
+func generateHashSides(commitmentProof *MerkleProof) ([]bool, error) {
+	pos := commitmentProof.LeafIndex
+	width := commitmentProof.NumberOfLeaves
+
+	if pos < width {
+		return nil, fmt.Errorf("leaf position %v is too high in proof with %v leaves", pos, width)
+	}
+
+	// TODO: this float casting is lossy, find a base 2 log function that operates on uint64 or assert that the width doesn't exceed 2**63-1
+	numSides := (int)(math.Ceil(math.Log2(float64(width))))
```

Comment:
You could probably try something like `math/bits` `LeadingZeros64` function. This lots like you are trying to find the most significant set bit.

https://go.dev/play/p/WaW0jcHilXt

Reply:
That's clever! 😮 We could use the index of the most significant bit, assuming the least significant bit is at index 0. We'd need to add the bitwise-and of the smaller bits to account for trees that aren't full, like we do with `math.Ceil`. But then we sidestep casting.

Reply:
Looks like res1 in the example should be `res1 := bits.LeadingZeros64(n - 1)`, but otherwise perfect. I'll put this in, thanks! 🙏